src/crypto/sha256.cpp

```diff
@@ -716,6 +716,60 @@ void CSHA256::Midstate(unsigned char hash[OUTPUT_SIZE], uint64_t* len, unsigned
     }
 }
 
+std::vector<unsigned char> CSHA256::Save() const {
+    size_t buf_size = bytes % 64;
+    std::vector<unsigned char> result(40 + buf_size);
+
+    WriteBE32(&result[ 0], s[0]);
+    WriteBE32(&result[ 4], s[1]);
+    WriteBE32(&result[ 8], s[2]);
+    WriteBE32(&result[12], s[3]);
+    WriteBE32(&result[16], s[4]);
+    WriteBE32(&result[20], s[5]);
+    WriteBE32(&result[24], s[6]);
+    WriteBE32(&result[28], s[7]);
+
+    WriteLE64(&result[32], bytes << 3);
+
+    memcpy(&result[40], buf, buf_size);
+
+    return result;
+}
+
+bool CSHA256::Load(const std::vector<unsigned char>& vch) {
+    if (vch.size() < 40) return false;
+
+    uint64_t bits = ReadLE64(&vch[32]);
+    size_t buf_size = (bits >> 3) % 64;
+
+    if ((bits & 0x07) != 0 || vch.size() != 40 + buf_size) return false;
+
+    // We want to leave the internal state of the object unchanged if false is returned.
+    // So no member variables can be modified until now.
+
+    s[0] = ReadBE32(&vch[ 0]);
+    s[1] = ReadBE32(&vch[ 4]);
+    s[2] = ReadBE32(&vch[ 8]);
+    s[3] = ReadBE32(&vch[12]);
+    s[4] = ReadBE32(&vch[16]);
+    s[5] = ReadBE32(&vch[20]);
+    s[6] = ReadBE32(&vch[24]);
+    s[7] = ReadBE32(&vch[28]);
+
+    bytes = bits >> 3;
+
+    memcpy(buf, &vch[40], buf_size);
+
+    return true;
+}
+
+bool CSHA256::SafeWrite(const unsigned char* data, size_t len) {
+  const uint64_t SHA256_MAX = 0x1FFFFFFFFFFFFFFF; // SHA256's maximum allowed message length in bytes.
+  if (SHA256_MAX < bytes || SHA256_MAX - bytes < len) return false;
```

Comment:
Checking for `SHA256_MAX - MAX_SCRIPT_ELEMENT_SIZE` in `Load()` would allow to get rid of three `if`s, because Write() will be always safe in regard to this check: the size of data on the stack cannot be bigger than `MAX_SCRIPT_ELEMENT_SIZE`. Why you think that doing this in `Write()` is better ? Note that in `OP_SHA256_INITIALIZE` it is even impossible to fail this check, because the `CTX` is new and can only increase up to `MAX_SCRIPT_ELEMENT_SIZE`. I think an `if` check is due in `Load()` and here this condition can just be checked with `assert` to be safe.

Reply:
Is it because you want to allow for `UPDATE` with user-supplied midstate to be able to go up to `SHA256_MAX` without any other limits ?

Reply:
You cold make `Load()` take the length and then check `SHA_MAX - bytes < len` inside `Load()`.
Although this makes Load/Save api less generic. But if you want it to be more generic so it can be used in other places, you probably want that check inside `Load()` anyway, because in general you cannot prevent anyone to use just `Write()` instead of `SafeWrite()` after `Load()`

Reply:
I suppose I could write `Load` to return a `int64_t` that returns the number of bytes used so far (or alternatively the number of remaining bytes available to be processed), and return `-1` for failure.

Is this really all worthwhile just to avoid one `if` statement in `SHA256INITALIZE`?

Reply:
I suppose we also get rid of `SafeWrite` which is kinda nice.

Reply:
all three `if` statement for `Write` is not needed if the `Write` is always safe,  if the check `SHA_MAX - bytes < len` is done inside Load() - at least in regard to manipulating the state from the script. The `Load()` can just fail if the subsequent `Write()` will be invalid. But my opinion is not strong on this - this is mostly bikeshedding. But I would like to add that check inside `Load()` would follow 'fail early' principle, which I favour.

Reply:
Having `Load` take in an argument of `the number of bytes I intend to write later` is such an unusual API, that I am very hesitant to implement it that way.

Reply:
Just call it `Prepare()` then :-)

Reply:
As I said, I have no strong opinion on this, and returning number of bytes available from `Load` and then checking it might be more messy than `SafeWrite` approach, because you will need to have that check in two places (in UPDATE and FINALIZE)